libtenzir/builtins/formats/cef.cpp

```diff
@@ -60,146 +62,122 @@ std::string unescape(std::string_view value) {
   return result;
 }
 
-/// A shallow representation a of a CEF message.
-struct message_view {
-  uint16_t cef_version;
-  std::string device_vendor;
-  std::string device_product;
-  std::string device_version;
-  std::string signature_id;
-  std::string name;
-  std::string severity;
-  record extension;
-};
-
 /// Parses the CEF extension field as a sequence of key-value pairs for further
 /// downstream processing.
 /// @param extension The string value of the extension field.
 /// @returns A vector of key-value pairs with properly unescaped values.
-caf::expected<record> parse_extension(std::string_view extension) {
-  record result;
-  auto splits = detail::split_escaped(extension, "=", "\\");
-  if (splits.size() < 2)
-    return caf::make_error(ec::parse_error, fmt::format("need at least one "
-                                                        "key=value pair: {}",
-                                                        extension));
-  // Process intermediate 'k0=a b c k1=d e f' extensions. The algorithm splits
-  // on '='. The first split is a key and the last split is a value. All
-  // intermediate splits are "reversed" in that they have the pattern 'a b c k1'
-  // where 'a b c' is the value from the previous key and 'k1`' is the key for
-  // the next value.
-  auto key = splits[0];
-  // Strip leading whitespace on first key. The spec says that trailing
-  // whitespace is considered part of the previous value, except for the last
-  // space that is split on.
-  for (size_t i = 0; i < key.size(); ++i)
-    if (key[i] != ' ') {
-      key = key.substr(i);
-      break;
+auto parse_extension(std::string_view extension,
+                     auto builder) -> std::optional<diagnostic> {
+  if (extension.empty()) {
+    return {};
+  }
+  auto key = extension.substr(0, extension.find('='));
+  extension.remove_prefix(key.size() + 1);
+  while (not extension.empty()) {
+    auto next_kv_sep = extension.find('=');
+    while (next_kv_sep != extension.npos
+           and extension[next_kv_sep - 1] == '\\') {
+      next_kv_sep = extension.find('=', next_kv_sep + 1);
+    }
```

Comment:
This also has an out-of-bounds read for `next_kv_sep == 0`. Since this is a recurring theme here, can you check _all_ handrolled parsers that you wrote for things like that?